skyportal/handlers/api/plan.py

```diff
@@ -0,0 +1,450 @@
+    maxtiles = plan_args['doMaxTiles']
+
+    params = params_struct(
+        localization.dateobs,
+        tobs=np.asarray(plan_args['tobs']),
+        filt=plan_args['filt'],
+        exposuretimes=exposuretimes,
+        probability=plan_args['probability'],
+        instrument=instrument,
+        schedule_type=plan_args['schedule_type'],
+        doReferences=plan_args['doReferences'],
+        doUsePrimary=plan_args['doUsePrimary'],
+        filterScheduleType=plan_args['filterScheduleType'],
+        schedule_strategy=plan_args['schedule_strategy'],
+        mindiff=plan_args['mindiff'],
+        doCompletedObservations=completed,
+        cobs=plan_args['cobs'],
+        doPlannedObservations=planned,
+        doMaxTiles=maxtiles,
+        max_nb_tiles=plan_args['max_nb_tiles'],
+        doRASlice=plan_args['doRASlice'],
+        raslice=plan_args['raslice'],
+        doBalanceExposure=plan_args['doBalanceExposure'],
+        airmass=plan_args['airmass'],
+    )
+
+    params['map_struct'] = dict(
+        zip(['prob', 'distmu', 'distsigma', 'distnorm'], localization.flat)
+    )
+
+    if plan_args['usePrevious']:
+        previous_telescope, previous_name = plan_args['previous_plan'].split("-")
+        plan_previous = ObservingPlan.query.filter_by(
+            dateobs=dateobs, telescope=previous_telescope, plan_name=previous_name
+        ).one()
+        ipix_previous = list(
+            {
+                i
+                for planned_observation in plan_previous.planned_observations
+                if planned_observation.field.ipix is not None
+                for i in planned_observation.field.ipix
+            }
+        )
+        params['map_struct']['prob'][ipix_previous] = 0.0
+
+    params['is3D'] = localization.is_3d
+    params['localization_name'] = localization.localization_name
+    map_struct, tile_structs, coverage_struct = gen_structs(params)
+
+    filter_ids = {"g": 1, "r": 2, "i": 3, "z": 4, "J": 5}
+    for ii in range(len(coverage_struct["ipix"])):
+        data = coverage_struct["data"][ii, :]
+        filt = coverage_struct["filters"][ii]
+        filter_id = filter_ids[filt]
+        mjd = data[2]
+        tt = time.Time(mjd, format='mjd')
+
+        overhead_per_exposure = params["config"][instrument.name][
+            "overhead_per_exposure"
+        ]
+
+        exposure_time, field_id, prob = data[4], data[5], data[6]
+
+        planned_observation = PlannedObservation(
+            obstime=tt.datetime,
+            dateobs=dateobs,
+            field_id=field_id,
+            exposure_time=exposure_time,
+            weight=prob,
+            filter_id=filter_id,
+            instrument_id=instrument.id,
+            planned_observation_id=ii,
+            plan_name=plan_name,
+            overhead_per_exposure=overhead_per_exposure,
+        )
+
+        DBSession().add(planned_observation)
+    plan.status = plan.Status.READY
+    DBSession().merge(plan)
+    request_handler.verify_and_commit()
```

Comment:
let's move this line out to the main handler method definition

Reply:
but how does that work if it is asynchronous?

Reply:
ah right, nm then